### PR TITLE
Feature/actor state polymorphism

### PR DESCRIPTION
**Pull request by wer1st**
### Description of the Change
Introduce universal actor states.
Get actor state from State Provider.
Move all concrete state logic to State Manager.
Move all common actor types to the new "types" folder.

### Benefits
One front state to work.

### CODE REVIEW

File: core/api/node_api.hpp [hunk 1]
```diff
@@ -140,7 +143,8 @@ namespace fc::api {
   };
 
   struct MinerPower {
-    Claim miner, total;
+    Claim miner;
+    Claim total;
```
Comment (turuslan) [inline core/api/node_api.hpp:147]:
```suggestion
    Claim total;
    bool has_min_power;
```
json name "HasMinPower"

File: core/vm/actor/builtin/types/miner/dead_line.hpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
+/**
```
Comment (turuslan) [inline core/vm/actor/builtin/types/miner/dead_line.hpp:1]: Please rename to `deadline.hpp`

File: core/vm/actor/builtin/states/state_provider.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    template <typename T, typename Tv0, typename Tv2, typename Tv3>
+    outcome::result<std::shared_ptr<T>> getCommonStatePtr(
+        const Actor &actor) const {
+      const auto version = getVersion(actor.code);
+
+      switch (version) {
+        case ActorVersion::kVersion0: {
+          OUTCOME_TRY(state, getStatePtr<Tv0>(actor.head));
+          return std::static_pointer_cast<T>(state);
```
Comment (turuslan) [inline core/vm/actor/builtin/states/state_provider.hpp:82]: Double free

File: core/vm/actor/builtin/states/state_provider.hpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        case ActorVersion::kVersion2: {
+          OUTCOME_TRY(state, getStatePtr<Tv2>(actor.head));
+          return std::static_pointer_cast<T>(state);
+        }
+        case ActorVersion::kVersion3: {
+          OUTCOME_TRY(state, getStatePtr<Tv3>(actor.head));
+          return std::static_pointer_cast<T>(state);
+        }
+      }
```
Comment (turuslan) [inline core/vm/actor/builtin/states/state_provider.hpp:92]:
```suggestion
      std::shared_ptr<T> state;
      switch (version) {
        case ActorVersion::kVersion0: {
          OUTCOME_TRYA(state, getStatePtr<Tv0>(actor.head));
          break;
        }
        case ActorVersion::kVersion2: {
          OUTCOME_TRYA(state, getStatePtr<Tv2>(actor.head));
          break;
        }
        case ActorVersion::kVersion3: {
          OUTCOME_TRYA(state, getStatePtr<Tv3>(actor.head));
          break;
        }
      }
      return state;
```

File: core/vm/actor/builtin/states/state_provider.hpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#include "vm/actor/builtin/states/reward_actor_state.hpp"
+#include "vm/actor/builtin/states/system_actor_state.hpp"
+#include "vm/actor/builtin/states/verified_registry_actor_state.hpp"
+
+namespace fc::vm::actor::builtin::states {
+
+  class StateProvider {
+   public:
+    explicit StateProvider(IpldPtr ipld);
+    ~StateProvider() = default;
```
Comment (turuslan) [inline core/vm/actor/builtin/states/state_provider.hpp:30]:
```suggestion
```

File: core/vm/actor/builtin/types/payment_channel/voucher.hpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+             && lane == rhs.lane && nonce == rhs.nonce && amount == rhs.amount
+             && min_close_height == rhs.min_close_height && merges == rhs.merges
+             && signature_bytes == rhs.signature_bytes;
+    }
+
+    inline outcome::result<Buffer> signingBytes() const {
+      auto copy = *this;
+      copy.signature_bytes = boost::none;
+      OUTCOME_TRY(signable_bytes, codec::cbor::encode(copy));
+      return std::move(signable_bytes);
```
Comment (turuslan) [inline core/vm/actor/builtin/types/payment_channel/voucher.hpp:82]:
```suggestion
      return codec::cbor::encode(copy);
```

File: core/vm/actor/builtin/types/payment_channel/voucher.hpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+             amount,
+             min_close_height,
+             merges,
+             signature_bytes)
+
+  struct PaymentVerifyParams {
+    Buffer extra;
+    Buffer proof;
+  };
+  CBOR_TUPLE(PaymentVerifyParams, extra, proof);
```
Comment (turuslan) [inline core/vm/actor/builtin/types/payment_channel/voucher.hpp:102]:
```suggestion
  CBOR_TUPLE(PaymentVerifyParams, extra, proof)
```

File: core/vm/actor/builtin/types/miner/dead_line.hpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+     * to charge extra gas for creation.
+     * @param ipld - ipld with empty adt::Array stored
+     * @param empty_amt_cid
+     */
+    inline static Deadline makeEmpty(IpldPtr ipld, const CID &empty_amt_cid) {
+      // TODO (a.chernyshov) initialize amt with correct bitwidth
+      // construct with empty already cid stored in ipld to avoid gas charge
+      Deadline deadline;
+      deadline.partitions = adt::Array<Partition>(empty_amt_cid, ipld);
+      deadline.expirations_epochs = adt::Array<RleBitset>(empty_amt_cid, ipld);
```
Comment (turuslan) [inline core/vm/actor/builtin/types/miner/dead_line.hpp:30]:
```suggestion
      deadline.partitions = {empty_amt_cid, ipld};
      deadline.expirations_epochs = {empty_amt_cid, ipld};
```
Can omit type name

**Review by Alexey-N-Chernyshov — COMMENTED** [review]
`StateProvider` is always used after `state_tree.get`, my suggestion is to make `StateProvider` the part of `StateTree`. The best option is to return actor state by address like `provider.getXXXActorState(address)` without resolving it with 'StateTree', but there is `checks.cpp` with remote interface with only 'StateGetActor' and remote ipfs.

File: core/vm/actor/builtin/states/state.hpp [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+#pragma once
+
+#include "vm/actor/actor.hpp"
+
+namespace fc::vm::actor::builtin::states {
+  struct State {
+    State(ActorType type, ActorVersion version)
+        : type(type), version(version) {}
+    virtual ~State() = default;
+
```
[the remark this answers is not in the document]
Comment (turuslan) [reply core/vm/actor/builtin/states/state.hpp:15]: Removing actor type and making `State` an abstract-virtual class sounds good

File: core/vm/actor/builtin/states/impl/state_manager_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    outcome::result<VerifiedRegistryActorStatePtr>
+    getVerifiedRegistryActorState() const override;
+
+    outcome::result<void> commitState(
+        const std::shared_ptr<State> &state) override;
+
+   private:
+    template <typename T, typename Tv0, typename Tv2, typename Tv3>
+    std::shared_ptr<T> createStatePtr(ActorVersion version) const {
+      std::shared_ptr<T> state;
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/builtin/states/impl/state_manager_impl.hpp:76]: `state` is not used.
```suggestion
```

File: core/vm/actor/builtin/v2/storage_power/storage_power_actor_utils.cpp [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+#include "vm/actor/builtin/v2/storage_power/storage_power_actor_utils.hpp"
+
+namespace fc::vm::actor::builtin::v2::storage_power {
+  outcome::result<void> PowerUtils::validateMinerHasClaim(
+      states::PowerActorStatePtr state, const Address &miner) const {
+    REQUIRE_NO_ERROR(state->loadClaimsRoot(), VMExitCode::kErrIllegalState);
+    REQUIRE_NO_ERROR_A(
+        has, state->hasClaim(miner), VMExitCode::kErrIllegalState);
```
Comment (turuslan) [inline core/vm/actor/builtin/v2/storage_power/storage_power_actor_utils.cpp:13]: Claims hamt was ([original](https://github.com/filecoin-project/cpp-filecoin/pull/344/files#diff-4c5fe7cd0bb07b6e8f829c5b7afc0ef72d0fccc72873b55ccc9445d77e35dd39L110-L111)) copied out from state on purpose.
New changes caused regression, when state with loaded hamt charges excessive gas on commit

File: core/vm/actor/builtin/states/state.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (wer1st) [reply core/vm/actor/builtin/states/state.hpp:15]: `virtual outcome::result<what??> getCbor() = 0`
we could not use  `outcome::result<State>`
we can use  outcome::result<shared_ptr<State>>, but in this case we should always make some cast after receiving this pointer

File: test/core/vm/actor/builtin/v0/init/init_actor_test.cpp [hunk 12]
```diff
@@ -27,6 +28,24 @@ namespace fc::vm::actor::builtin::v0::init {
...
+
+      EXPECT_CALL(*state_manager, getInitActorState())
+          .Times(testing::AnyNumber())
+          .WillRepeatedly(testing::Invoke([&]() {
+            EXPECT_OUTCOME_TRUE(cid, ipld->setCbor(state));
+            EXPECT_OUTCOME_TRUE(current_state,
+                                ipld->getCbor<InitActorState>(cid));
+            auto s = std::make_shared<InitActorState>(current_state);
+            return std::static_pointer_cast<states::InitActorState>(s);
+          }));
```
[the remark this answers is not in the document]
Comment (wer1st) [reply test/core/vm/actor/builtin/v0/init/init_actor_test.cpp:48]: Also we have some tests that are not inherited from the common fixture. I suggest to create another ticket to optimise actor tests. What do you think?

File: core/api/node_api.hpp (hunk 1 above)
Comment (wer1st) [reply core/api/node_api.hpp:147]: It turns out that this is not necessary

**Review by turuslan — APPROVED** [review]

**Review by turuslan — APPROVED** [review]

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**